src/Libraries/OrchardCore.Commerce.Abstractions/Models/OrderPart.cs

```diff
@@ -28,14 +27,9 @@ public class OrderPart : ContentPart
     /// <summary>
     /// Gets the amounts charged for this order. Typically a single credit card charge.
     /// </summary>
-
-    // This is a temporary solution, it needs to be reworked in the future!
-#pragma warning disable CA2326 // Do not use TypeNameHandling values other than None
-#pragma warning disable SCS0028 // TypeNameHandling is set to the other value than 'None'. It may lead to deserialization vulnerability.
-    [JsonProperty(ItemTypeNameHandling = TypeNameHandling.Auto)]
-#pragma warning restore SCS0028 // TypeNameHandling is set to the other value than 'None'. It may lead to deserialization vulnerability.
-#pragma warning restore CA2326 // Do not use TypeNameHandling values other than None
-    public IList<IPayment> Charges { get; } = new List<IPayment>();
+    // Due to the significant intentional limitations of polymorphic deserialization in System.Text.Json this type now
+    // only supports Payment instead of any arbitrary IPayment implementation.
```

Comment:
This was almost 2 months ago, so I had to retrace my steps and do some new research. I guess I misunderstood something. MS [intentionally didn't implement this feature in .NET 6](https://github.com/dotnet/runtime/issues/30969#issuecomment-535779492) and [it's still contrived to use since .NET 7](https://learn.microsoft.com/en-us/dotnet/standard/serialization/system-text-json/polymorphism?pivots=dotnet-8-0#polymorphic-type-discriminators), but it can be made to work. Indeed [OC has a DI solution for it](https://docs.orchardcore.net/en/latest/releases/2.0.0/#:~:text=Any%20serializable%20object%20that%20contains%20a%20polymorphic%20property%20(a%20base%20type%20that%20can%20contain%20sub%2Dclasses%20instances)%20needs%20to%20register%20all%20possible%20sub%2Dclasses%20this%20way:).

That said, I actually want to double down because I never liked the whole `IPayment` thing. I guess the idea was that different payment providers may return custom payment models? I think this is pointless and convoluted. This is very different from how OC uses polymorphic deserialization, e.g. for admin menu nodes, which is way more justified.

A new major release is the perfect time to get rid of the `IPayment` interface, and simply use a single common `Payment` model everywhere. afaik nobody has made custom payment providers yet (partly because we haven't [documented how to do it](https://github.com/OrchardCMS/OrchardCore.Commerce/issues/448) yet) so it shouldn't be intrusive.

Reply:
I lack the necessary level of understanding of OCC to comment, so let's not have `IPayment`, then. However, we need to obviously keep the ability to use multiple payment providers, a third-party one being implemented in the same way as the ones built into OCC. Will this still be supported?

BTW you might take inspiration from how Mike solved something similar under https://github.com/OrchardCMS/OrchardCore/pull/16402: there, `Query` is not polymorphic, but to support storing data for arbitrary Query types (like SQL Queries), it's an `Entity`. Thus it can store arbitrary data in its `Properties`, which are done deserialized ad-hoc with `As<T>()` (see e.g. how `SqlQueryMetadata` is used).

Reply:
> However, we need to obviously keep the ability to use multiple payment providers, a third-party one being implemented in the same way as the ones built into OCC. Will this still be supported?

Yes, of course.

> BTW you might take inspiration from how Mike solved something similar

Thanks, it's interesting.